### ISSUE TITLE
Provide an `associate` terminal operators

### ISSUE THREAD

**Issue by jcornaz** [Kotlin/kotlinx.coroutines#1541]
Hi, 

Today we realized that despite the terminal operators `toList` and `toSet` being provided for `Flow`, there isn't anything to create a map from a `Flow`.

Could we have `associate`, `associateBy` and `associateWith` for `Flow`, as we have for `Sequence`?

**Comment by qwwdfsad** [comment]
Definitely, we can implement these operators.

Initially, we've tried to avoid having the same variety of Flow operators as Sequence has because Flow has slightly different use-cases though it looks like a sequence. But if people keep lacking these operators for Flow, we will add them.

**Comment by fvasco** [comment]
Please consider `Flow<Pair<K, V>>.toMap()`, too.

**Comment by Globegitter** [comment]
Yep, right now we have to workaround by doing `.toList().toMap()` - would be nice to be able to convert to a map out of the box.

**Comment by Xaseron** [comment]
I'have came here because i find the `toList().toMap()` workaround kinda ugly.
Should we create a separate issue?

**Comment by elizarov** [comment]
Please, describe your specific use-cases in this issue.

**Comment by Xaseron** [comment]
i have functions that returns a `Flow<Pair<K, V>>`. 
To convert this into a Map i have to use `toList().toMap()`.
Having a terminal operator `toMap` on flow would elimante an intermediate collection created by `toList` and thus increase the performance.

**Comment by fvasco** [comment]
@Xaseron,
you should to explain a bit deeper your use case.

However you can consider to include an extension function like:

```kotlin
suspend fun <K, V> Flow<Pair<K, V>>.toMap(): Map<K, V> {
    val map = mutableMapOf<K, V>()
    collect { (k, v) ->
        map[k] = v
    }
    return map
}
```

**Comment by Xaseron** [comment]
Having this in the stdlib would make it even with collection api. 
Isn't this a goal?

**Comment by elizarov** [comment]
@Xaseron So, what's your use-case? Can you elaborate, please.

**Comment by Xaseron** [comment]
I have this method where i use a Mongodb Aggregation Pipeline to gather all values for specific year/month in a list. Afterwards i'm calling collect on this publisher to map each year/month to the average value of this list and emit it.

```kotlin
    fun overallTrends(metricName: String) = flow {
        val mongoList = col.aggregate<OverallTrend>(
                """
[
    { $ group:
        {
             _id: {
                year: "$ metaData.reportingYear",
                month: "$ metaData.reportingMonth"
            },
            values: { $ push: "$ $metricName" }
        }
    }
]""".formatJson()
        )
        mongoList.collect { trend ->
            val yearMonth = trend.id.toYearMonth()
            emit(yearMonth to trend.averageValue)
        }
    }
```
I want to call a terminal operation on this flow to get a Map. 
Therefore currently i have to it this way:
`val overallTrendsMap = mongoMetricsService.overallTrends(mongoMetricName).toList().toMap()`
I would be more performant and nicer to read if i could directly call it like this:
`val overallTrendsMap = mongoMetricsService.overallTrends(mongoMetricName).toMap()`

**Comment by elizarov** [comment]
Why do you define your `overallTrends` function to return a `flow` if you plan to get get a map out of it? Wouldn't it be easier to define it a suspending function that returns a map in the first place?

**Comment by clamothe** [comment]
My use case is using an asynchronous Mongo driver, performing a find operation that returns a reactive Publisher. I convert this into a Flow. Some consumers of this Flow need to create a map out of the results

**Comment by Inego** [comment]
Our use-case is for `groupBy`, not `associateBy`, but I believe it also fits in here:
We have to consume a gRPC server stream method from a coroutine. We wrap it in a `callbackFlow` and then have to group values from that flow. Ideally there should be a standard terminal operation like `groupBy` which exists in Collections, but since there is none, we either have to resort to collection of an intermediary list and then grouping it, or to use `fold` which is boilerplatey:

```kotlin
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.fold
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.runBlocking

data class Row(
    val ticker: String,
    val trdAcc: String
)

fun main() {

    // ... Comes from a callbackFlow-wrapped gRPC stream method call
    val rows = flowOf(
        Row("T", "a"),
        Row("T", "b"),
        Row("U", "c")
    )

    runBlocking {
        
        // Prepare a map of tickers and their trade accounts.

        // Option 1:
        // Undesirable because of intermediary collection to a list
        val option1: Map<String, List<String>> = rows.toList()
            .groupBy({ it.ticker }) { it.trdAcc }

        // Option 2:
        // Wordy because it's essentially what StdLib's `groupBy` does
        val option2: Map<String, List<String>> = rows.fold(LinkedHashMap<String, MutableList<String>>()) { acc, row ->
            acc.apply {
                getOrPut(row.ticker) { ArrayList() }.add(row.trdAcc)
            }
        }
    }
}

```

**Comment by dkhalanskyjb** [comment]
Sorry to say I've completely forgotten about this issue, but yes, we will implement all the functions requested here. The reasoning that `Sequence` has an operator is enough for us now to provide it for `Flow` as well: https://github.com/Kotlin/kotlinx.coroutines/issues/4212#issuecomment-2286323663.
